File: Parser.cs
using ChatReciver.utils;
using Newtonsoft.Json.Linq;
using System.Text.RegularExpressions;
using System.Linq;

namespace ChatReciver
{
    internal class Parser
    {
        //Done
        public static LivePageData GetOptionsFromLivePage(string resp, string id)
        {
#pragma warning disable CS8604 // Possible null reference argument.
            LivePageData data = new()
            {
                liveId = id
            };

            string api = @"['""]INNERTUBE_API_KEY['""]:\s*['""](.+?)['""]";
            string cliver = @"['""]clientVersion['""]:\s*['""]([\d.]+?)['""]";
            string cont = @"['""]continuation['""]:\s*['""](.+?)['""]";

            Match mApi = Regex.Match(resp, api);
            Match mCliVer = Regex.Match(resp, cliver);
            Match mCont = Regex.Match(resp, cont);

            if (!mApi.Success)
            {
                throw new Exception("Unable to find api key");
            }
            else
            {
                data.apiKey = JObject.Parse("{" + mApi.Value + "}")
                                     .GetValue("INNERTUBE_API_KEY")
                                     .Value<string>();
            }

            if (!mCliVer.Success)
            {
                throw new Exception("Unable to find client version");
            }
            else
            {
                data.clientVersion = JObject.Parse("{" + mCliVer.Value + "}")
                                     .GetValue("clientVersion")
                                     .Value<string>();
            }

            if (!mCont.Success)
            {
                throw new Exception("Unable to find client version");
            }
            else
            {
                data.continuation = JObject.Parse("{" + mCont.Value + "}")
                                     .GetValue("continuation")
                                     .Value<string>();
            }

            return data;
#pragma warning restore CS8604 // Possible null reference argument.
        }

        public static ChatData ParseChatData(JObject json)
        {

#pragma warning disable CS8604 // Possible null reference argument.
            if (json == null) throw new Exception("Response unavaliable");

            ChatData data = new();

            JArray actions = json["continuationContents"]?["liveChatContinuation"]?["actions"]?.Value<JArray>();
            ChatItem[] objectArray = new ChatItem[actions.Count];

            if (actions != null)
            {
                int index = 0;
                foreach (JObject item in actions.Cast<JObject>())
                {
                    if (!item.ContainsKey("addChatItemAction")) continue;
                    if (!item["addChatItemAction"].Contains("item")) continue;
                    if (!item["addChatItemAction"]["item"].Contains("liveChatTextMessageRenderer")) continue;
                    objectArray[index] = ParseActionToChatItem(item);
                    index++;
                }
            }

            data.chatItems = objectArray;
            JObject continuationData = json["continuationContents"]["liveChatContinuation"]["continuations"][0].Value<JObject>();
            string continuation = "";

            if (continuationData["invalidationContinuationData"] != null)
            {
                continuation = continuationData["invalidationContinuationData"]["continuation"].Value<String>();
            }
            else if (continuationData["timedContinuationData"] != null)
            {
                continuation = continuationData["timedContinuationData"]["continuation"].Value<String>();
            }

            data.continuation = continuation;

#pragma warning restore CS8604 // Possible null reference argument.
            return data;
        }


        public static string ParseThumbnail(JArray data)
        {
            JObject imageData = data.Last.Value<JObject>();
            return imageData["url"].Value<string>();
        }

        public static string ConverColorToHex6(uint colorNum)
        {
            return "#" + (colorNum.ToString("X"))[..].ToUpper();
        }

        public static string ParseMessage(JArray messages)
        {
            string[] strings = new string[messages.Count];
            int i = 0;
            foreach (JObject item in messages.Cast<JObject>())
            {
                if (messages.Contains("text"))
                {
                    strings[i] = item["text"].Value<string>();
                } else
                {

                    strings[i] = item["emoji"]["emojiId"].Value<string>();
                }
                i++;
            }
            string output = string.Join("; ", strings);
            return output;
        }

        public static JObject RenderFromAction(JObject data) 
        {
            if (data["addChatItemAction"] == null) return null;
            JObject item = data["addChatItemAction"]["item"].Value<JObject>();

            if (IsNotNull(item["liveChatTextMessageRenderer"])) return item["liveChatTextMessageRenderer"].Value<JObject>();
            else if (IsNotNull(item["liveChatPaidMessageRenderer"])) return item["liveChatPaidMessageRenderer"].Value<JObject>();
            else if (IsNotNull(item["liveChatPaidStickerRenderer"])) return item["liveChatPaidStickerRenderer"].Value<JObject>();
            else if (IsNotNull(item["liveChatMembershipItemRenderer"])) return item["liveChatMembershipItemRenderer"].Value<JObject>();
            else return null;
        }

        private static bool IsNotNull(JToken? jToken)
        {
            return jToken != null;
        }

        public static ChatItem ParseActionToChatItem(JObject data)
        {
            Console.WriteLine(data.ToString());
            ChatItem item = new ();
            JObject messageRender = RenderFromAction(data);
            if (messageRender == null) return null;

            JArray messages = new();

            if(data.ContainsKey("message"))
            {
                messages = data["message"]?["runs"]?.Value<JArray>();
            } else if (data.ContainsKey("headerSubtext"))
            {
                messages = data["headerSubtext"]["runs"].Value<JArray>();
            }
            string authorNameText = messageRender["authorName"]?["simpleText"]?.Value<string>() ?? "";

            item.id = messageRender["id"].Value<string>();

            Author author = new()
            {
                name = authorNameText,
                thumbnail = ParseThumbnail(messageRender["authorPhoto"]["thumbnails"].Value<JArray>()),
                channelID = messageRender["authorExternalChannelId"].Value<string>()
            };
            if (messageRender.ContainsKey("authorBadges"))
            {
                foreach (var entry in from JObject entry in messageRender["authorBadges"].Value<JArray>()
                                      where !entry.ContainsKey("customThumbnail")
                                      select entry)
                {

                    //switch (entry["liveChatAuthorBadgeRenderer"]["icon"]["iconType"].Value<string>())
                    //{
                    //    case "OWNER":
                    //        author.isOwner = true;
                    //        break;
                    //    case "VERIFIED":
                    //        author.isVerified = true;
                    //        break;
                    //    case "MODERATOR":
                    //        author.isModerator = true;
                    //        break;
                    //}
                }
            }

            item.author = author;
            item.message = ParseMessage(messages);
            item.timestamp = DateTime.UnixEpoch.AddMicroseconds(double.Parse(messageRender["timestampUsec"].Value<string>()));
            if(messageRender.ContainsKey("sticker"))
            {
                Superchat superchat = new()
                {
                    amount = messageRender["stickerpurchaseAmountText"]["simpleText"].Value<string>(),
                    color = ConverColorToHex6(uint.Parse(messageRender["backgroundColor"].Value<string>())),
                    sticker = ParseThumbnail(messageRender["sticker"]["thumbnails"].Value<JArray>())
                };
                item.superchat = superchat;
            } else if (messageRender.ContainsKey("purchaseAmountText"))
            {
                Superchat superchat = new()
                {
                    amount = messageRender["purchaseAmountText"]["simpleText"].Value<string>(),
                    color = ConverColorToHex6(uint.Parse(messageRender["bodyBackgroundColor"].Value<string>()))
                };
                item.superchat = superchat;
            }
            return item;
        }
    }
}

File: api/src/main/java/cz/iqlandia/iqplanetarium/chatreciever/Parser.java
package cz.iqlandia.iqplanetarium.chatreciever;

import cz.iqlandia.iqplanetarium.chatreciever.records.ChatData;
import cz.iqlandia.iqplanetarium.chatreciever.records.ChatItem;
import org.json.JSONArray;
import org.json.JSONObject;

import java.awt.*;
import java.text.ParseException;
import java.time.Instant;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.List;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Parser {
    public static LivePageData getLivePageData(String data) throws ParseException {
        String liveID, apiKey, clientVersion, continuation;

        Pattern pattern = Pattern.compile("<link rel=\"canonical\" href=\"https://www.youtube.com/watch\\?v=(.+?)\">");
        Matcher matcher = pattern.matcher(data);

        if (matcher.find()) {
            liveID = matcher.group(1);
        } else {
            throw new ParseException("Unable to find liveID!", 0);
        }

        pattern = Pattern.compile("['\"]isReplay['\"]:\\s*(true)");
        matcher = pattern.matcher(data);

        if(matcher.find()) {
            throw new ParseException("This video is a replay!", matcher.start(0));
        }

        pattern = Pattern.compile("['\"]INNERTUBE_API_KEY['\"]:\\s*['\"](.+?)['\"]");
        matcher = pattern.matcher(data);

        if (matcher.find()) {
            apiKey = matcher.group(1);
        } else {
            throw new ParseException("Unable to find apiKey!", 0);
        }

        pattern = Pattern.compile("['\"]clientVersion['\"]:\\s*['\"]([\\d.]+?)['\"]");
        matcher = pattern.matcher(data);

        if (matcher.find()) {
            clientVersion = matcher.group(1);
        } else {
            throw new ParseException("Unable to find clientVersion!", 0);
        }

        pattern = Pattern.compile("['\"]continuation['\"]:\\s*['\"](.+?)['\"]");
        matcher = pattern.matcher(data);

        if (matcher.find()) {
            continuation = matcher.group(1);
        } else {
            throw new ParseException("Unable to find continuation!", 0);
        }

        return new LivePageData(liveID,apiKey,clientVersion,continuation);
        //return null;
    }

    public static ChatData parseChatData(JSONObject data) {
        List<ChatItem> chatItems = new ArrayList<>();
        if (data.getJSONObject("continuationContents").getJSONObject("liveChatContinuation").has("actions")) {
            if (data.getJSONObject("continuationContents")
                    .getJSONObject("liveChatContinuation")
                    .getJSONArray("actions") != null) {
                chatItems = JSONToList(data.getJSONObject("continuationContents").getJSONObject("liveChatContinuation").getJSONArray("actions")).stream()
                        .map(o -> (JSONObject) o)
                        .map(Parser::parseActionToChatItem)
                        .filter(Objects::nonNull).toList();
            }
        }

        JSONObject continuationData = data.getJSONObject("continuationContents")
                .getJSONObject("liveChatContinuation")
                .getJSONArray("continuations")
                .getJSONObject(0);

        String continuation = "";

        if(continuationData.has("invalidationContinuationData")) {
            continuation = continuationData
                    .getJSONObject("invalidationContinuationData")
                    .getString("continuation");
        } else if (continuationData.has("timedContinuationData")) {
            continuation = continuationData
                    .getJSONObject("timedContinuationData")
                    .getString("continuation");
        }

        return new ChatData(chatItems, continuation);
    }

    private static List<Object> JSONToList(JSONArray list) {
        List<Object> objects = new ArrayList<>();
        for (Object o : list) {
            objects.add(o);
        }
        return objects;
    }

    public static ChatItem.ImageItem parseThumbnailToImageItem(List<String> data, String alt) {
        String url = data.get(data.size() - 1);
        if(url != null) {
            return new ChatItem.ImageItem(url, alt);
        } else return new ChatItem.ImageItem("","");
    }

    public static Color convertColorToHex6(int colorNum) {
        return new Color(colorNum,true);
    }

    public static String parseMessages(List<String> runs) {
        StringJoiner js = new StringJoiner("");
        for (String run : runs) {
            js.add(run);
        }
        return js.toString();
    }

    private static ChatItem parseActionToChatItem(JSONObject action) {
        if(!action.has("addChatItemAction")) return null;

        JSONObject metaItem = action.getJSONObject("addChatItemAction").getJSONObject("item");

        if(!metaItem.has("liveChatTextMessageRenderer")) {
            return null;
        }

        JSONObject item = metaItem.getJSONObject("liveChatTextMessageRenderer");

        JSONArray messages = item.getJSONObject("message").getJSONArray("runs");
        String message = parseMessages(JSONToList(messages).stream()
                .map(o -> (JSONObject) o)
                .map(Parser::convertToString)
                .filter(Objects::nonNull)
                .toList());

        String authorName = item.has("authorName")?item.getJSONObject("authorName").getString("simpleText") : "<unable to parse>";
        String authorID = item.getString("authorExternalChannelId");

        ChatItem.ImageItem thumbnail = parseThumbnailToImageItem(JSONToList(item.getJSONObject("authorPhoto").getJSONArray("thumbnails")).stream()
                .map((o) -> (JSONObject) o)
                .map(o -> o.getString("url"))
                .filter(Objects::nonNull)
                .toList(),authorName);

        Optional<ChatItem.Author.Badge> badge = Optional.empty();
        boolean isVerified = false, isOwner = false, isModerator = false;

        if(item.has("authorBadges")) {
            for (JSONObject o : JSONToList(item.getJSONArray("authorBadges")).stream().map(o -> (JSONObject) o).toList()) {
                if (o.has("customThumbnail")) {
                    badge = Optional.of(new ChatItem.Author.Badge(
                            parseThumbnailToImageItem(JSONToList(o.getJSONObject("customThumbnail").getJSONArray("thumbnails")).stream()
                                    .map((ob) -> (JSONObject) ob)
                                    .map(ob -> ob.getString("url"))
                                    .filter(Objects::nonNull)
                                    .toList(),o.getString("tooltip")),
                            o.getString("tooltip")
                    ));
                } else if (o.has("icon")){
                    switch (o.getJSONObject("icon").getString("iconType")) {
                        case "OWNER" -> isOwner = true;
                        case "VERIFIED" -> isVerified = true;
                        case "MODERATOR" -> isModerator = true;
                    }
                }
            }
        }


        ChatItem.Author author = new ChatItem.Author(authorName, Optional.of(thumbnail), authorID, badge, isVerified, isOwner, isModerator);
        return new ChatItem(item.getString("id"), author, message, ZonedDateTime.ofInstant(Instant.ofEpochSecond(item.getLong("timestampUsec")/1000000), ZoneId.of("Europe/Prague")));
    }

    private static String convertToString(JSONObject o) {
        if(o.has("text")) {
            return o.getString("text");
        } else if (o.has("emoji")) {
            if(o.getJSONObject("emoji").has("isCustomEmoji")) {
                return ":"+o.getJSONObject("emoji")
                        .getJSONObject("image")
                        .getJSONObject("accessibility")
                        .getJSONObject("accessibilityData")
                        .getString("label") + ":";
            } else {
                return o.getJSONObject("emoji")
                        .getJSONObject("image")
                        .getJSONObject("accessibility")
                        .getJSONObject("accessibilityData")
                        .getString("label");
            }
        }
        else return null;
    }
}
